tempesta_fw/http_parser.c

```diff
@@ -1495,7 +1477,69 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 		}
 	}
 
-	RGEN_LWS();
+	/* In request we should pass empty headers:
+	 * RFC 7230 5.4:
+	 * ....
+	 * ....
+	 * If the authority component is missing or
+	 * undefined for the target URI, then a client MUST send a Host header
+	 * field with an empty field-value.
+	 */
+	__FSM_STATE(RGen_LWS) {
+		switch (c) {
+		case '\r':
+			TFW_DBG3("\\r symbol\n");
+			if (parser->_header_tag >= 0) {
+				/* For some headers, like Host, we should add
+				 * possibility of empty value. For such headers,
+				 * we set parser->_header_tag. If this field
+				 * is set, and header is empty, we call 
+				 * tfw_http_msg_hdr_close for it. So, this
+				 * field has 2 purposes: as flag, that empty
+				 * header is possible and shows, with which
+				 * header we work at the moment.
+				 */
+				if (likely(!(parser->flags & TFW_HTTP_PF_CRLF))) {
+					TFW_DBG3("CR and LF flag not set. Setting CR.\n");
+					parser->flags |= TFW_HTTP_PF_CR;
+				} else {
+					TFW_DBG3("CR or LF already set\n");
+				}
+				tfw_http_msg_hdr_chunk_fixup(msg, data, p - data);
+				if (tfw_http_msg_hdr_close(msg, parser->_header_tag))
+					return TFW_BLOCK;
+				parser->_header_tag = -1;
+				__FSM_MOVE(RGen_LF);
+			} else {
+				if (likely(!(parser->flags & TFW_HTTP_PF_LF))) {
```

Comment:
This appears to be a bug. This block of code should be the same as in the original macro/state.